contracts/Tokens/Prime/PrimeLiquidityProvider.sol

```diff
@@ -0,0 +1,340 @@
+     * @param distributionSpeeds_ New distribution speeds for tokens
+     * @custom:access Controlled by ACM
+     * @custom:error Throw InvalidArguments on different length of tokens and speeds array
+     */
+    function setTokensDistributionSpeed(address[] calldata tokens_, uint256[] calldata distributionSpeeds_) external {
+        _checkAccessAllowed("setTokensDistributionSpeed(address[],uint256[])");
+        uint256 numTokens = tokens_.length;
+
+        if (numTokens != distributionSpeeds_.length) {
+            revert InvalidArguments();
+        }
+
+        for (uint256 i; i < numTokens; ) {
+            _setTokenDistributionSpeed(tokens_[i], distributionSpeeds_[i]);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /**
+     * @notice Claim all the token accrued till last block
+     * @param token_ The list of tokens to claim tokens
+     * @custom:event Emits TokenTransferredToPrime event
+     * @custom:error Throw InvalidArguments on Zero address(token)
+     */
+    function releaseFunds(address token_) external {
+        if (fundsTransferpaused) {
+            revert FundsTransferIspaused();
+        }
+
+        accrueTokens(token_);
+        uint256 accruedAmount = tokenAmountAccrued[token_];
+        tokenAmountAccrued[token_] = 0;
+
+        IERC20Upgradeable(token_).safeTransfer(prime, accruedAmount);
+
+        emit TokenTransferredToPrime(token_, accruedAmount);
+    }
+
+    /**
+     * @notice A public function to sweep accidental ERC-20 transfers to this contract. Tokens are sent to user
+     * @param token_ The address of the ERC-20 token to sweep
+     * @param to_ The address of the recipient
+     * @param amount_ The amount of tokens needs to transfer
+     * @custom:event Emits SweepToken event
+     * @custom:error Throw InsufficientBalance on Zero address(token)
+     * @custom:access Only Governance
+     */
+    function sweepToken(IERC20Upgradeable token_, address to_, uint256 amount_) external onlyOwner {
+        uint256 balance = token_.balanceOf(address(this));
+        if (amount_ > balance) {
+            revert InsufficientBalance(amount_, balance);
+        }
+
+        token_.safeTransfer(to_, balance);
+
+        emit SweepToken(address(token_), to_, amount_);
+    }
+
+    /**
+     * @notice Accrue token by updating the distribution state
+     * @param token_ Address of the token
+     * @custom:event Emits TokensAccrued event
+     */
+    function accrueTokens(address token_) public {
+        _ensureZeroAddress(token_);
+
+        uint256 blockNumber = getBlockNumber();
+        uint256 deltaBlocks = blockNumber - lastAccruedBlock[token_];
+
+        if (deltaBlocks > 0) {
+            uint256 distributionSpeed = tokenDistributionSpeeds[token_];
+            uint256 balance = IERC20Upgradeable(token_).balanceOf(address(this));
+            uint256 tokenAccrued;
+
+            if (distributionSpeed > 0 && balance > 0) {
```

Comment:
We can be more restrictive here, to ensure we still have some balance available to distribute:
```suggestion
            if (distributionSpeed > 0 && (balance - tokenAmountAccrued[token_]) > 0) {
```